### PR TITLE
qt: upstream Linux changes

### PR DESCRIPTION
**Pull request by danielnachun**
- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [ ] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

File: Formula/qt.rb
Comment (carlocab) [inline Formula/qt.rb:105]:
```suggestion
      config_args << "-sysroot" << MacOS.sdk_path.to_s
```
Push this later; we've already got a bunch of long timeout jobs running currently.

**Comment by carlocab** [comment]
```
In file included from /tmp/qt-20210828-20615-1neorpd/qt-everywhere-src-6.1.2/qtquick3d/src/plugins/assetimporters/assimp/../../../3rdparty/assimp/src/include/assimp/GenericProperty.h:52,
                 from /tmp/qt-20210828-20615-1neorpd/qt-everywhere-src-6.1.2/qtquick3d/src/3rdparty/assimp/src/code/Common/BaseProcess.h:46,
                 from /tmp/qt-20210828-20615-1neorpd/qt-everywhere-src-6.1.2/qtquick3d/src/3rdparty/assimp/src/code/Common/BaseProcess.cpp:44:
/tmp/qt-20210828-20615-1neorpd/qt-everywhere-src-6.1.2/qtquick3d/src/plugins/assetimporters/assimp/../../../3rdparty/assimp/src/include/assimp/Importer.hpp:505:16: error: 'exception_ptr' in namespace 'std' does not name a type; did you mean 'exception'?
  505 |     const std::exception_ptr& GetException() const;
      |                ^~~~~~~~~~~~~
      |                exception
```

**Comment by danielnachun** [comment]
There is an upstream commit I had to apply to fix that failure and I got it to build locally.  Like with `qt@5`, I had to change `g++-base.conf` to use GCC 11 for the test to pass.  I hit a new, unrelated problem in the last part of the test that I haven't been able to fix.  I think I can push the fixes I have so far so we have a bottle at least for other maintainers to download and try to work with.

**Comment by danielnachun** [comment]
The bottle is now built!  The current test failure we have has been very frustrating to deal with locally.  I added `libheif` as a Linux dependency and in several builds I test this worked.  However, it has started failing now.  I'm suspecting a possible race condition where something is being built too soon and doesn't pick up`libheif`.  I am going to locally test `ENV.deparallelize` to see if that fixes it.  This would make the build process even more slower than it currently is, but that may be the only way to make it work.

**Comment by cho-m** [comment]
Does Qt support HEIF on Linux? 

Since Apple is the main adopter of HEIF, it makes sense that the image format is available on macOS. I see the following directory in source code `qtimageformats/src/plugins/imageformats/macheif`.

From quick search, I didn't find anything else outside mac-specific plugin.

**Comment by cho-m** [comment]
If HEIF is not supported on Linux, then we can drop dependency and modify test to ignore on Linux:
```cpp
        for(const char* fmt:{"bmp", "cur", "gif",
      #ifdef __APPLE__
          "heic", "heif",
      #endif
          "icns", "ico", "jp2", "jpeg", "jpg", "pbm", "pgm", "png",
```

When I tried locally, this got past the first testcase; however, I hit another error on next testcase. It tried to use `/home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libGL.so` rather than `/home/linuxbrew/.linuxbrew/Cellar/mesa/21.2.1/lib/libGL.so`
```
g++-11 -Wl,-O1 -Wl,-rpath,/home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib -o test main.o   /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libQt6Svg.so /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libQt6Widgets.so /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libQt63DCore.so /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libQt6Quick3D.so /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libQt6Quick3DRuntimeRender.so /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libQt6Quick.so /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libQt6OpenGL.so /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libQt6Quick3DAssetImport.so /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libQt6Quick3DUtils.so /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libQt6ShaderTools.so /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libQt6Gui.so /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libGL.so /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libQt6NetworkAuth.so /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libQt6QmlModels.so /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libQt6Qml.so /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libQt6Network.so /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libQt6Sql.so /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libQt6Core.so -lpthread -lGL
/home/linuxbrew/.linuxbrew/bin/ld: cannot find /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libGL.so: No such file or directory
collect2: error: ld returned 1 exit status
Makefile:254: recipe for target 'test' failed
make: *** [test] Error 1
```

**Comment by danielnachun** [comment]
> If HEIF is not supported on Linux, then we can drop dependency and modify test to ignore on Linux:
> 
> ```c++
>         for(const char* fmt:{"bmp", "cur", "gif",
>       #ifdef __APPLE__
>           "heic", "heif",
>       #endif
>           "icns", "ico", "jp2", "jpeg", "jpg", "pbm", "pgm", "png",
> ```

I think this is right call.  Although we do have a `libheif` which provides Linux support for HEIF, I see that Arch for example does not use `libheif` as a dependency for QT 6.  So it seems this may be a macOS only image format for now.

> When I tried locally, this got past the first testcase; however, I hit another error on next testcase. It tried to use `/home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libGL.so` rather than `/home/linuxbrew/.linuxbrew/Cellar/mesa/21.2.1/lib/libGL.so`
> 
> ```
> g++-11 -Wl,-O1 -Wl,-rpath,/home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib -o test main.o   /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libQt6Svg.so /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libQt6Widgets.so /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libQt63DCore.so /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libQt6Quick3D.so /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libQt6Quick3DRuntimeRender.so /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libQt6Quick.so /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libQt6OpenGL.so /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libQt6Quick3DAssetImport.so /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libQt6Quick3DUtils.so /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libQt6ShaderTools.so /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libQt6Gui.so /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libGL.so /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libQt6NetworkAuth.so /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libQt6QmlModels.so /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libQt6Qml.so /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libQt6Network.so /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libQt6Sql.so /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libQt6Core.so -lpthread -lGL
> /home/linuxbrew/.linuxbrew/bin/ld: cannot find /home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libGL.so: No such file or directory
> collect2: error: ld returned 1 exit status
> Makefile:254: recipe for target 'test' failed
> make: *** [test] Error 1
> ```

I also got stuck with this error.  It does not make a lot of sense because `-lGL` is also added, which should be finding the library dynamically.  So why does it also try to specify an exact location for `libGL` as well?

**Comment by cho-m** [comment]
The GL library path is saved inside some `*.prl` file, like `lib/libQt6Gui.prl`:
```
QMAKE_PRL_BUILD_DIR = /tmp/qt-20210829-20628-1rf57n8/qt-everywhere-src-6.1.2/qtbase/src/gui
QMAKE_PRL_TARGET = libQt6Gui.so
QMAKE_PRL_CONFIG = shared
QMAKE_PRL_VERSION = 6.1.2
QMAKE_PRL_LIBS = $$[QT_INSTALL_LIBS]/libQt6Core.so -lpthread $$[QT_INSTALL_LIBS]/libGL.so
QMAKE_PRL_LIBS_FOR_CMAKE = $$[QT_INSTALL_LIBS]/libQt6Core.so;-lpthread;$$[QT_INSTALL_LIBS]/libGL.so
```

Manually changing libGL path in the `*.prl` files fixes failure.

We should try to find if there is a build option to modify this.

**Comment by danielnachun** [comment]
This is definitely a bug: https://bugreports.qt.io/browse/QTBUG-85801.  I'm going to see if this commit fixes it: https://invent.kde.org/qt/qt/qtbase/-/commit/9c0f448f5a4e2e96e9b801c751c0a145c0eb6dad.patch.

**Comment by danielnachun** [comment]
It looks like that commit was already applied, so that isn't the problem.

**Comment by cho-m** [comment]
There are some other related problems in `*.prl` files, like in `lib/libQt6InputSupport.prl`:
`$$[QT_INSTALL_LIBS]/libxkbcommon.so -L/home/linuxbrew/.linuxbrew/Cellar/systemd/246/lib`

The same problem will occur for `libxkbcommon`.
On the other hand, `systemd` library can be found, but will break on any version update since it is using versioned Cellar path.

**Comment by danielnachun** [comment]
I think we should open a bug report on this upstream.  I can see in the output from the test posted above that `-lGL`, for example, has been correctly added the linker flags.  The problem here is that the `/home/linuxbrew/.linuxbrew/Cellar/qt/6.1.2/lib/libGL.so` is either not being deleted, or being added incorrectly.  If it wasn't there, `-lGL` would be found in `$HOMEBREW_PREFIX/lib` where `libGL.so` is linked, and the test would work fine.

**Comment by cho-m** [comment]
The main thing to figure out is how the `*.prl` files are created. These appear to be used by `qmake` and are the reason for current issues.

It would be good if Qt provides a way to updating these during build, like how paths are added to `*.cmake` and `*.pc` files.

At runtime, it may be possible to override `QT_INSTALL_LIBS` to `#{HOMEBREW_PREFIX}/lib`, but, even if it works, it isn't a user-friendly approach.

The last-ditch workaround would be to manually modify the files and either delete these paths or point to correct paths.

**Comment by danielnachun** [comment]
There is already a CMake file which is used to update the files after the build: https://invent.kde.org/qt/qt/qtbase/-/blob/dev/cmake/QtFinishPrlFile.cmake.  They previously made some changes that were supposed to address the issue of using incorrect prefixes for library paths, but it seems those fixes were incomplete.

**Comment by MikeMcQuaid** [comment]
Have resolved conflicts here and updated based on https://github.com/Homebrew/brew/pull/11955

**Comment by MikeMcQuaid** [comment]
@danielnachun I'd be tempted to punt on HEIF support on Linux for now to get this PR merged in some form.

**Comment by danielnachun** [comment]
I'm totally okay with skipping HEIF support on Linux for now.  Unfortunately the issue with invalid library paths in PRL files is more serious, as I believe it would mean that any formula which uses `qmake` would be broken.  I've spent enough time digging through the source tree and commit history of Qt 6 to be convinced that this is a genuine bug.  I think the only way forward to is to open a bug report with upstream and manually remove the invalid paths with `inreplace`.  We'll have to make a list of all the affected PRL files, and another list of all the problematic library paths, and iterate over those to make sure they are removed.

**Comment by MikeMcQuaid** [comment]
> I'm totally okay with skipping HEIF support on Linux for now.

👍🏻 

> Unfortunately the issue with invalid library paths in PRL files is more serious, as I believe it would mean that any formula which uses `qmake` would be broken.

How does this differ from e.g. the linuxbrew-core version? Is this a Linux-only bug?

**Comment by cho-m** [comment]
> How does this differ from e.g. the linuxbrew-core version? Is this a Linux-only bug?

Based on quick glance, I don't think Linuxbrew has ever bottled Qt 6. The upstream Qt 6 PR (https://github.com/Homebrew/linuxbrew-core/pull/22412) was opened when Homebrew changed `qt` from v5 to v6

It seems to be Linux-only as macOS uses system frameworks, e.g.
```
QMAKE_PRL_LIBS = -F$$[QT_INSTALL_LIBS] -framework QtCore -framework DiskArbitration -framework IOKit -framework OpenGL -framework AGL -framework AppKit -framework ImageIO -framework Metal
QMAKE_PRL_LIBS_FOR_CMAKE = -F$$[QT_INSTALL_LIBS];-framework;QtCore;-framework DiskArbitration;-framework IOKit;-framework OpenGL;-framework AGL;-framework AppKit;-framework ImageIO;-framework Metal
```

---

May be worth looking further into `QtFinishPrlFile.cmake` and seeing if there is some logic that doesn't work with Homebrew/Linuxbrew's packaging and/or directory structure.

**Comment by cho-m** [comment]
Some notes:
1. Homebrew formula sets `std_cmake_args(install_prefix: HOMEBREW_PREFIX` , so `CMAKE_INSTALL_PREFIX == HOMEBREW_PREFIX`
2. If `QT_BUILD_INTERNALS_RELOCATABLE_INSTALL_PREFIX`, is not defined, it is set to `CMAKE_INSTALL_PREFIX` (https://invent.kde.org/search?search=qt_get_relocatable_install_prefix&group_id=1000&project_id=1216&scope=&search_code=true&snippets=false&repository_ref=dev&nav_source=navbar)
3. Qt adds `QT_BUILD_INTERNALS_RELOCATABLE_INSTALL_PREFIX/lib` to `QT_LIB_DIRS`, https://invent.kde.org/qt/qt/qtbase/-/blob/dev/cmake/QtPrlHelpers.cmake#L136-138
    ```cmake
        if(QT_WILL_INSTALL)
            list(APPEND qt_lib_dirs
                 "${QT_BUILD_INTERNALS_RELOCATABLE_INSTALL_PREFIX}/${INSTALL_LIBDIR}")
    ```
4. Qt will consider any library that is in same directory or a subdirectory of `QT_LIB_DIRS` as a Qt library: https://invent.kde.org/qt/qt/qtbase/-/blob/dev/cmake/QtGenerateLibHelpers.cmake#L105-122

**Comment by danielnachun** [comment]
@MikeMcQuaid as @cho-m said we could not bottle this in linuxbrew-core because of disk space issues.  I wasn't sure where to put this, but we shouldn't hold up the migration process while trying to get this bottled.

@iMichka I think we might want to align the linuxbrew-core version of this with the current version in homebrew-core, so there is no diff.  Then when we fix this you can merge it.

@cho-m I will try setting `QT_BUILD_INTERNALS_RELOCATABLE_INSTALL_PREFIX` to `Formula["qt"].opt_lib` so that it doesn't mistakenly think that `libGL.so` or other brewed libraries from other formulae are QT formulae.  As for the `systemd` library incorrectly using the cellar path instead of the `Formula["systems"].opt_lib`, we might just use an `inreplace` for now unless we can find a variable to override the path.

**Comment by MikeMcQuaid** [comment]
> @MikeMcQuaid as @cho-m said we could not bottle this in linuxbrew-core because of disk space issues. I wasn't sure where to put this, but we shouldn't hold up the migration process while trying to get this bottled.
> 
> @iMichka I think we might want to align the linuxbrew-core version of this with the current version in homebrew-core, so there is no diff. Then when we fix this you can merge it.

👍🏻 works for me.

**Comment by danielnachun** [comment]
I added `-DQT_BUILD_INTERNALS_RELOCATABLE_INSTALL_PREFIX=#{prefix}` to the `cmake` args and it stopped adding the incorrect library paths!  Really great job hunting that down @cho-m, I will add a comment to this addition to explain this for future reference.

As I had suspected, that fix does not adress the issue with the cellar path being used for `systemd` instead of the `opt_lib`.  I think for this one we will have to use an `inreplace`.  I believe the following code would achieve that, but if anyone has improvements, please let me know:

```
if OS.linux?
  Pathname.glob("#{lib}/*.prl") do |prl|
    if File.readlines(prl).grep(Formula["systemd"].lib)
      inreplace prl, Formula["systemd"].lib, Formula["systemd"].opt_lib
    end
  end
end
```

I'm not sure if I need `Regexp.escape` for `Formula["systems"].lib`.

**Comment by danielnachun** [comment]
I've pushed the fixes that I think will make this pass the tests on Linux now.  Once we've figured out the right `inreplace` for they `systemd` cellar paths, I think this will be good to go.

**Comment by cho-m** [comment]
I wonder if there is a way to avoid all the dependencies' versioned Cellar paths inside bottle. This may be due to how Homebrew's `pkg-config` files usually return the Cellar path.

Inside `lib/*.prl` files, these include:
```
@@HOMEBREW_CELLAR@@/libxkbcommon/1.3.1/lib
@@HOMEBREW_CELLAR@@/systemd/246/lib
@@HOMEBREW_CELLAR@@/hunspell/1.7.0_2/lib
```

Inside `share/**/*.pri` files, most formulae are seen like:
```
@@HOMEBREW_CELLAR@@/dbus/1.12.20
@@HOMEBREW_CELLAR@@/systemd/246
@@HOMEBREW_CELLAR@@/libb2/0.98.1
@@HOMEBREW_CELLAR@@/brotli/1.0.9
@@HOMEBREW_CELLAR@@/libxau/1.0.9
@@HOMEBREW_CELLAR@@/libxdmcp/1.1.3
...
```

There is also a versioned CMake reference in `bin/qt-cmake`:
```
original_cmake_path="@@HOMEBREW_CELLAR@@/cmake/3.21.2/bin/cmake"
```
But this at least falls back to `cmake` command, so whatever is on PATH.

File: Formula/qt.rb
Comment (carlocab) [inline Formula/qt.rb:169]: Why change the code after wrapping it in `if OS.mac?`

File: Formula/qt.rb
Comment (carlocab) [inline Formula/qt.rb:150]:
```suggestion
    inreplace lib/"cmake/Qt6/qt.toolchain.cmake" do |s|
      s.gsub! %r{#{HOMEBREW_SHIMS_PATH}/[^/]+/super/#{ENV.cc}}, ENV.cc
      s.gsub! %r{#{HOMEBREW_SHIMS_PATH}/[^/]+/super/#{Regexp.escape(ENV.cxx)}}, ENV.cxx if OS.linux?
    end
```
We could probably also use `Superenv.shims_path`, but we'll then need to wait for #84863.

File: Formula/qt.rb
Comment (danielnachun) [reply Formula/qt.rb:169]: Whoops, I think this got bungled because of a merge conflict from another PR that I didn't fully handle right.  I will push the fix for this along with any other requested changes.

File: Formula/qt.rb
Comment (danielnachun) [reply Formula/qt.rb:150]: Yes, I was going to comment that this shim replacement seems like a good candidate for a new DSL since it's a fairly common fix we have to make.  I'll make the change you suggested for now.  I'm inclined to say we shouldn't wait for https://github.com/Homebrew/homebrew-core/pull/84863 if that would be the only thing holding this up, but we should do whatever is least painful in terms of CI usage.

**Comment by carlocab** [comment]
> ```
> if OS.linux?
>   Pathname.glob("#{lib}/*.prl") do |prl|
>     if File.readlines(prl).grep(Formula["systemd"].lib)
>       inreplace prl, Formula["systemd"].lib, Formula["systemd"].opt_lib
>     end
>   end
> end
> ```

```ruby
if OS.linux?
  systemd = Formula["systemd"]
  lib.glob("*.prl") do |prl|
    if prl.read.include? systemd.lib.realpath
      inreplace prl, systemd.lib.realpath, systemd.opt_lib
    end
  end
end
```
or
```ruby
if OS.linux?
  systemd = Formula["systemd"]
  lib.glob("*.prl") do |prl|
    # Passing `false` prevents errors from files where there is nothing to replace
    inreplace prl, systemd.lib.realpath, systemd.opt_lib, false
  end
end
```

**Comment by danielnachun** [comment]
> **EDIT:** Some of these may still work as long as the dependency formula is linked as it should still be able to find `-l` libraries even when `-L` path doesn't exist, as long as `HOMEBREW_PREFIX/lib` is searched by linker. Similar situation with `-I` headers in linked path.
> 
> Anything that is keg_only or unlinked can cause issues for users who want to use `qmake` when the version/revision is bumped. Most are currently linked by default, though, there was the idea of making `systemd` keg_only

I think you are right that this comes from how `pkg-config` returns prefixes and I'm not really sure how to fix this without using `inreplace`.  Some formulae may use `qmake` to build and this could create issues if they try to look in directories that no longer exist when the dependencies are updated (and we do not want to rebuild `qt` after every dependency update).  

Given that `systemd` isn't the only dependency with this issue, I'm thinking we can just iterate over all the dependencies for `qt`, and replace any cellar paths as needed.  Here is roughly what that might look like:

```ruby
if OS.linux?
  # Replace cellar paths with opt paths so that qmake
  # doesn't break after dependency updates
  cmake_files = [
    bin/"qt-cmake",
    bin/"qt-cmake-private",
  ]
  cmake = Formula["cmake"]
  cmake_files do |cmake_file|
    if cmake_file.read.include? cmake.bin.realpath
      inreplace cmake_file, cmake.bin.realpath, cmake.opt_bin
    end
  end
  
  qt_deps = Formula["qt"].deps
  prl_files = lib.glob("*.prl")
  prl_files do |prl_file|
    qt_deps do |dep|
      if prl_file.read.include? dep.lib.realpath
        inreplace prl_file, dep.lib.realpath, dep.opt_lib
      end
    end
  end

  pri_files = (share/"qt/mkspecs/modules").glob("*.pri")
  pri_files << share/"qt/mkspecs/qmodule.pri" 
  pri_files do |pri_file|
    qt_deps do |dep|
      if pri_file.read.include? dep.include.realpath
        inreplace pri_file, dep.include.realpath, dep.opt_include
      end
    end
  end

end
```

Any suggestions to improve this (or if this is even the right approach) would be great!

**Comment by MikeMcQuaid** [comment]
> I added `-DQT_BUILD_INTERNALS_RELOCATABLE_INSTALL_PREFIX=#{prefix}` to the `cmake` args and it stopped adding the incorrect library paths! Really great job hunting that down @cho-m, I will add a comment to this addition to explain this for future reference.

Nice work both!

> I wonder if there is a way to avoid all the dependencies' versioned Cellar paths inside bottle. This may be due to how Homebrew's `pkg-config` files usually return the Cellar path.

This makes me wonder if we should/could just fixup `.pc` files when they are installed and/or bottled?

> ```ruby
>       if file.read.include? dep.lib.realpath
>         inreplace file, dep.lib.realpath, dep.opt_lib
>       end
> ```

My only issue with this pattern is that this logic will stick around indefinitely even if there's no matches rather than having `inreplace` error out (which is why it does so when there's no matches). Similarly, you're reading the file every time through the loop and probably want to only load it once.

**Comment by danielnachun** [comment]
> This makes me wonder if we should/could just fixup `.pc` files when they are installed and/or bottled?

In that case would you suggest we merge without trying to fix (I still need to make some other unrelated changes first), and fix up these files whenever the relevant formulae are updated?  We'd need to rebottle Qt afterward because the PRL/PRI files would still have the incorrect cellar path.

> My only issue with this pattern is that this logic will stick around indefinitely even if there's no matches rather than having `inreplace` error out (which is why it does so when there's no matches). Similarly, you're reading the file every time through the loop and probably want to only load it once.

Yeah this is not an ideal solution - if we decide to fix the PC files in the other formulae then we wouldn't need to do this at all.

**Comment by MikeMcQuaid** [comment]
> In that case would you suggest we merge without trying to fix (I still need to make some other unrelated changes first), and fix up these files whenever the relevant formulae are updated? We'd need to rebottle Qt afterward because the PRL/PRI files would still have the incorrect cellar path.

I don't have strong views here. I do feel anything holding up the migration would be nice to be done sooner rather than later.

> Yeah this is not an ideal solution - if we decide to fix the PC files in the other formulae then we wouldn't need to do this at all.

👍🏻

File: Formula/qt.rb
Comment (danielnachun) [inline Formula/qt.rb:137]: I think it's preferable here to just remove `Superenv.shims_path`, rather than hardcode `/usr/bin`.

**Comment by danielnachun** [comment]
I think this is ready to go now.  I can add a list of the formulae whose `.pc` files need to be fixed so we have a record of that.

**Review by MikeMcQuaid — APPROVED** [review]
Greater work here @danielnachun!

**Comment by paperchalice** [comment]
> I wonder if there is a way to avoid all the dependencies' versioned Cellar paths inside bottle. This may be due to how Homebrew's `pkg-config` files usually return the Cellar path.

`pkg-config` has an option `--define-prefix`
> try to override the value of prefix for each .pc file found with a guesstimated value based on the location of the .pc file

Output from `brew install -s -i <some formula>`:
```
$ pkg-config libpkgconf --cflags --libs
-I/usr/local/Cellar/pkgconf/1.8.0/include/pkgconf -L/usr/local/Cellar/pkgconf/1.8.0/lib -lpkgconf

$ pkg-config libpkgconf --cflags --libs --define-prefix
-I/usr/local/opt/pkgconf/include/pkgconf -L/usr/local/opt/pkgconf/lib -lpkgconf
```
I'm not sure whether this option is meaningful in this situation.

**Pull request closed without merging**
